src/dev/lupio/lupio_tty.cc

```diff
@@ -0,0 +1,176 @@
+#include "dev/lupio/lupio_tty.hh"
+
+#include "base/trace.hh"
+#include "debug/LupioTTY.hh"
+#include "dev/platform.hh"
+#include "mem/packet.hh"
+#include "mem/packet_access.hh"
+#include "params/LupioTTY.hh"
+
+#define LUPIO_TTY_INVAL    0x80000000
+
+/* Same fields for CTRL and STAT registers */
+#define LUPIO_TTY_WBIT  (1 << 0)
+#define LUPIO_TTY_RBIT  (1 << 1)
+
+namespace gem5
+{
+
+/*
+ * IRQ management
+ */
+void
+LupioTTY::lupio_tty_update_irq()
+{
+    unsigned int irq;
+
+    irq = (writ_ie && writ_char != -1)
+        || (read_ie && read_char != -1);
+    if (irq) {
+        DPRINTF(LupioTTY, "LupioTTY InterEvent, interrupting\n");
+        platform->postConsoleInt();
+    } else {
+//        platform->clearConsoleInt();
+        DPRINTF(LupioTTY, "LupioTTY InterEvent, not interrupting\n");
+    }
+}
+
+void
+LupioTTY::dataAvailable()
+{
+      // unsure about these checks
+      if (read_ie && read_char == -1) {
+          DPRINTF(LupioTTY, "Data available, interrupting\n");
+          platform->postConsoleInt();
+     }
+}
+
+LupioTTY::LupioTTY(const Params &params)
+    : BasicPioDevice(params, params.pio_size), device(params.device),
+    byteOrder(params.byte_order),
+    platform(params.platform)
+{
+    // initialize internal registers
+    writ_char = read_char = -1;
+    writ_ie = read_ie = false;
+
+    // setup serial device callbacks
+    device->regInterfaceCallback([this]() { dataAvailable(); });
+}
+
+uint64_t
+LupioTTY::lupioTTYRead(uint8_t addr)
+{
+    uint32_t r = LUPIO_TTY_INVAL;
+    if (device->dataAvailable()){
+        if (read_char != -1)
+            fprintf(stderr, "Dropping characters\n");
+        else {
+            read_char = device->readData();     //read from the terminal
+            lupio_tty_update_irq();
+        }
+    }
+
+    switch (addr >> 2) {
+        case LUPIO_TTY_WRIT:
+            if (writ_char != -1) {
+                DPRINTF(LupioTTY, "Accessing write register\n");
+                r = writ_char;
```

Comment:
What is `r`? This variable name is too short (and it's unclear if it's a local variable or class variable).

Reply:
I think r stands for return value? I can definitely change this naming but we may also need to change it in the RNG and RTC since I think they also use "r".

Reply:
FWIW, `r` is the variable name typically used in QEMU for returning a value in devices' read function. It probably stands for "return value" or "register". That's the name I also used in my QEMU models. If it doesn't fit gem5's coding style, we could use something like `ret`.